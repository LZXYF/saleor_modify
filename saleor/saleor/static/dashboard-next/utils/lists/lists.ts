type List<TData> = TData[];
type Compare<TData> = (a: TData, b: TData) => boolean;

export function isSelected<TData>(
  data: TData,
  list: List<TData>,
  compare: Compare<TData>
) {
  return !!list.find(listElement => compare(listElement, data));
}

export function add<TData>(data: TData, list: List<TData>) {
  return [...list, data];
}

export function remove<TData>(
  data: TData,
  list: List<TData>,
  compare: Compare<TData>
) {
  return list.filter(listElement => !compare(listElement, data));
}

export function toggle<TData>(
  data: TData,
  list: List<TData>,
  compare: Compare<TData>
) {
  return isSelected(data, list, compare)
    ? remove(data, list, compare)
    : add(data, list);
}
