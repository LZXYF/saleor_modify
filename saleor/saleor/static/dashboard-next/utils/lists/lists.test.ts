import { add, isSelected, remove, toggle } from "./lists";

const initialArray = ["lorem", "ipsum", "dolor"];

describe("Properly calculates output arrays", () => {
  it("Adds", () => {
    expect(add("sit", initialArray)).toMatchSnapshot();
  });

  it("Removes", () => {
    expect(remove("ipsum", initialArray, (a, b) => a === b)).toMatchSnapshot();
  });

  it("Matches", () => {
    expect(isSelected("lorem", initialArray, (a, b) => a === b)).toBe(true);
    expect(isSelected("sit", initialArray, (a, b) => a === b)).toBe(false);
  });

  it("Toggles", () => {
    expect(toggle("lorem", initialArray, (a, b) => a === b)).toMatchSnapshot();
    expect(toggle("sit", initialArray, (a, b) => a === b)).toMatchSnapshot();
  });
});
