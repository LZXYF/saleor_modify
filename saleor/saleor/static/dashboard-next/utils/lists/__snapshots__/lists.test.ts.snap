// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Properly calculates output arrays Adds 1`] = `
Array [
  "lorem",
  "ipsum",
  "dolor",
  "sit",
]
`;

exports[`Properly calculates output arrays Removes 1`] = `
Array [
  "lorem",
  "dolor",
]
`;

exports[`Properly calculates output arrays Toggles 1`] = `
Array [
  "ipsum",
  "dolor",
]
`;

exports[`Properly calculates output arrays Toggles 2`] = `
Array [
  "lorem",
  "ipsum",
  "dolor",
  "sit",
]
`;
